util/conditions/experimental/merge_strategies.go

```diff
@@ -0,0 +1,408 @@
+type MergeStrategy interface {
+	// Merge all the condition in input.
+	//
+	// It is up to the caller to ensure that all the expected conditions exist (e.g. by adding new conditions with status Unknown).
+	// Condition in input must be of the given conditionTypes (other condition types must be discarded);
+	//
+	// The list of conditionTypes has an implicit order; it is up to the implementation of merge to use this info or not.
+	// If negativeConditionTypes are in scope, the implementation of merge should treat them accordingly.
+	//
+	// It required, the implementation of merge must add a stepCounter to the output message.
+	Merge(conditions []ConditionWithOwnerInfo, conditionTypes []string, negativeConditionTypes sets.Set[string], stepCounter bool) (status metav1.ConditionStatus, reason, message string, err error)
+}
+
+// ConditionWithOwnerInfo is a wrapper  metav1.Condition with addition ConditionOwnerInfo.
+// Those info can be used when generating the message resulting from the merge operation.
+type ConditionWithOwnerInfo struct {
+	OwnerResource ConditionOwnerInfo
+	metav1.Condition
+}
+
+// ConditionOwnerInfo contains info about the object that owns the condition.
+type ConditionOwnerInfo struct {
+	metav1.TypeMeta
+	Namespace string
+	Name      string
+}
+
+// defaultMergeStrategy defines the default merge strategy for Cluster API conditions.
+type defaultMergeStrategy struct{}
+
+type mergePriority uint8
+
+const (
+	issueMergePriority mergePriority = iota
+	unknownMergePriority
+	infoMergePriority
+)
+
+func newDefaultMergeStrategy() MergeStrategy {
+	return &defaultMergeStrategy{}
+}
+
+// Merge all the condition in input based on a strategy that surfaces issue first, then unknown conditions, the info (if none of issues and unknown condition exists).
+// - issues: conditions with positive polarity (normal True) and status False or conditions with negative polarity (normal False) and status True.
+// - unknown: conditions with status unknown.
+// - info: conditions with positive polarity (normal True) and status True or conditions with negative polarity (normal False) and status False.
+func (d *defaultMergeStrategy) Merge(conditions []ConditionWithOwnerInfo, conditionTypes []string, negativeConditionTypes sets.Set[string], stepCounter bool) (status metav1.ConditionStatus, reason, message string, err error) {
+	// Infer which operation is calling this func, so it is possible to use different strategies for computing the message for the target condition.
+	// - When merge should consider a single condition type, we can assume this func is called within an aggregate operation
+	//   (Aggregate should merge the same condition across many objects)
+	isAggregateOperation := len(conditionTypes) == 1
+
+	// - Otherwise we can assume this func is called within an summary operation
+	//   (Aggregate should merge the same condition across many objects)
```

Comment:
nit: Should this line explain what a summary operation does (vs. restating the same as l.93)?

Reply:
fixed